### DISCUSSION TITLE
Why In DokanDispatchWrite has condition to use irp->userbuffer?

### DISCUSSION THREAD

**Discussion by ylgybbz** [dokan-dev/dokany#1190] [Q&A]
In  DokanDispatchWrite,  there is code:
...
   if (RequestContext->Irp->MdlAddress) {
      buffer =
          MmGetSystemAddressForMdlNormalSafe(RequestContext->Irp->MdlAddress);
    } else {
      buffer = RequestContext->Irp->UserBuffer;
    }
...

Any Condition will use Irp->UserBuffer? I think the the volDeviceObject->Flags has set DO_DIRECT_IO in DokanMountVolume,  maybe it always use mdladdress?

**Comment by Liryna** [comment]
The `Irp->UserBuffer` was originially supported by Dokan legacy
https://github.com/dokan-dev/dokany/commit/192b852ed1ac58c448a884311f1ab8b671411e26 then later `DO_DIRECT_IO` was added by https://github.com/dokan-dev/dokany/commit/a815463467b8b08405b8a55b69704e544b61c9cf and probably comes from
https://github.com/microsoft/Windows-driver-samples/blob/7b0c7e24905cf4cafd33da5737b573696beb66b7/general/event/wdm/event.c#L169-L172

If we had to do something here, I tend to say that we could remove `DO_DIRECT_IO` but I am also afraid it breaks something I am missing
